workflows/methylation/methylation-cohort.wdl

```diff
@@ -0,0 +1,361 @@
+
+            return args
+
+        if __name__ == "__main__":
+            args = get_args()
+
+            # Read beta values and compute standard deviation
+            data = []
+            with open("beta.csv", "r") as f:
+                reader = csv.reader(f)
+                header = next(reader)
+                for i, line in enumerate(reader):
+                    probe = line[0]
+                    sd = np.std([float(x) for x in line[1:]])
+                    data.append([probe, sd])
+
+            sd_df = pd.DataFrame(data, columns=["probe", "sd"]).set_index("probe")
+
+            # Filter probes based on standard deviation
+            filtered_probes = sd_df.sort_values("sd", ascending=False).head(args.num_probes).index
+
+            pd.Series(filtered_probes, index=filtered_probes).to_csv('filtered_probes.csv', index=False)
+
+            # Filter beta values
+            with open("beta.csv", "r") as f:
+                reader = csv.reader(f)
+                header = next(reader)
+                header[0] = "probe"
+                beta_data = [line for line in reader if line[0] in filtered_probes]
+
+            bd = pd.DataFrame(beta_data, columns=header).set_index("probe")
+            bd.to_csv("filtered_beta.csv")
+        SCRIPT
+
+        python filter.py --num_probes ~{num_probes} beta.csv
+    >>>
+
+    output {
+        File filtered_beta_values = "filtered_beta.csv"
+        File filtered_probes = "filtered_probes.csv"
+    }
+
+    #@ except: ContainerValue
+    runtime {
+        container: "quay.io/biocontainers/pandas:2.2.1"
+        memory: "28 GB"
+        cpu: 1
+        disks: "~{disk_size_gb} GB"
+        maxRetries: 1
+    }
+}
+
+task generate_umap {
+    meta {
+        description: "Generate UMAP embedding"
+        outputs: {
+            umap: "UMAP embedding for all samples"
+        }
+    }
+
+    parameter_meta {
+        filtered_beta_values: "Filtered beta values for all samples"
+    }
+
+    input {
+        File filtered_beta_values
+    }
+
+    Int disk_size_gb = ceil(size(filtered_beta_values, "GiB") * 2)
+
+    #@ except: LineWidth
+    command <<<
+        ln -s ~{filtered_beta_values} beta.csv
+
+        python <<SCRIPT
+        import pandas as pd
+        import umap
+
+        # Read beta values
+        beta = pd.read_csv("beta.csv", index_col=0)
+
+        # Perform UMAP
+        embedding = umap.UMAP().fit_transform(beta.T)
+
+        # Save UMAP embedding
+        umap = pd.DataFrame(data=embedding, index=beta.T.index, columns=["UMAP1", "UMAP2"])
+        umap.index_name = "sample"
+        umap.to_csv("umap.csv")
+        SCRIPT
+    >>>
+
+    output {
+        File umap = "umap.csv"
+    }
+
+    #@ except: ContainerValue
+    runtime {
+        container: "ghcr.io/stjudecloud/umap:branch-methylation-0.5.7-1"
+        memory: "8 GB"
+        cpu: 1
+        disks: "~{disk_size_gb} GB"
+        maxRetries: 1
+    }
+}
+
+task plot_umap {
```

Comment:
This task is primarily for testing purposes. I'm not sure we want to plot the UMAPs here. They'll be hosted on Pecan and use the ProteinPaint scatter plot library.

Reply:
I think it would be useful (if not necessary) for non-Pecan users of the workflow. It is also quite a lightweight task, so I'm not concerned about wasted compute. Maybe a boolean toggle for skipping this step?